File: DiaryNutrition/app/src/main/java/ediger/diarynutrition/settings/SettingsFragment.java
package ediger.diarynutrition.settings;

import android.Manifest;
import android.app.AlarmManager;
import android.app.DatePickerDialog;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Environment;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.view.View;
import android.widget.DatePicker;
import android.widget.Toast;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;


import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.work.WorkInfo;

import com.google.common.util.concurrent.ListenableFuture;

import ediger.diarynutrition.Consts;
import ediger.diarynutrition.MainActivity;
import ediger.diarynutrition.PreferenceHelper;
import ediger.diarynutrition.R;
import ediger.diarynutrition.intro.PolicyActivity;
import ediger.diarynutrition.util.NutritionProgramUtils;
import ediger.diarynutrition.util.SnackbarUtils;

public class SettingsFragment extends PreferenceFragmentCompat implements SharedPreferences.
        OnSharedPreferenceChangeListener, DatePickerDialog.OnDateSetListener {

    public static final String KEY_PREF_DATA_LANGUAGE= "data_language";
    public static final String KEY_PREF_DATA_PATH = "data_path";
    public static final String KEY_PREF_DATA_BACKUP = "data_backup";
    public static final String KEY_PREF_DATA_RESTORE = "data_restore";
    public static final String KEY_PREF_POLICY = "policy";

    private static final String KEY_PREF_PROGRAM_EDIT = "program_edit";
    private static final String KEY_PREF_PROGRAM_RESET = "program_reset";
    private static final String KEY_PREF_PROGRAM_MEALS = "program_meals";

    private static final int REQ_WRITE_STORAGE = 112;

    private SettingsViewModel viewModel;

    private Calendar birthday = Calendar.getInstance();
    private SimpleDateFormat dateFormat = new SimpleDateFormat("d MMMM yyyy", Locale.getDefault());
    private Date date = new Date();

    // TODO: Get rid of this approach
    private boolean isBackupRequested = false;
    private boolean isRestoreRequested = false;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {

        setPreferencesFromResource(R.xml.preferences, rootKey);

        PreferenceHelper.getPreferences().registerOnSharedPreferenceChangeListener(this);

        viewModel = new ViewModelProvider(requireActivity()).get(SettingsViewModel.class);

        Preference heightPref = findPreference(Consts.KEY_HEIGHT);
        heightPref.setSummary(PreferenceHelper.getValue(Consts.KEY_HEIGHT, String.class, ""));

        Preference birthdayPref = findPreference(Consts.KEY_BIRTHDAY);
        date.setTime(PreferenceHelper.getValue(Consts.KEY_BIRTHDAY, Long.class, 0L));
        birthdayPref.setSummary(dateFormat.format(date));
        birthdayPref.setOnPreferenceClickListener(preference -> {
            showDateDialog();
            return false;
        });

        float cal = PreferenceHelper.getValue(Consts.KEY_PROGRAM_CAL, Float.class, 0f);
        Preference caloriesPref = findPreference(Consts.KEY_PROGRAM_CAL);
        caloriesPref.setSummary(String.valueOf((int) cal));
        caloriesPref.setOnPreferenceClickListener(preference -> {
            new ChangeCaloriesDialog().show(getParentFragmentManager(), null);
            return false;
        });

        Preference waterPref = findPreference(Consts.KEY_PROGRAM_WATER);
        waterPref.setSummary(String.valueOf(PreferenceHelper.getValue(
                Consts.KEY_PROGRAM_WATER,
                Integer.class,
                NutritionProgramUtils.getDefaultWater()
        )));
        waterPref.setOnPreferenceClickListener(preference -> {
            new ChangeWaterDialog().show(getParentFragmentManager(), null);
            return false;
        });

        Preference editProgram = findPreference(KEY_PREF_PROGRAM_EDIT);
        editProgram.setOnPreferenceClickListener(preference -> {
            startActivity(new Intent(requireActivity(), ProgramActivity.class));
            return false;
        });

        findPreference(KEY_PREF_PROGRAM_MEALS).setOnPreferenceClickListener(preference -> {
            NavHostFragment.findNavController(this)
                    .navigate(R.id.action_setting_to_user_meal);
            return false;
        });

        Preference resetProgram = findPreference(KEY_PREF_PROGRAM_RESET);
        resetProgram.setOnPreferenceClickListener(preference -> {
            NutritionProgramUtils.setToDefault();
            return false;
        });

        // TODO: Fix wrong path (Do I need it?)
        Preference pathData = findPreference(KEY_PREF_DATA_PATH);
        String path = Environment.getExternalStorageDirectory().getPath() +
                "/DiaryNutrition/" + Consts.DATABASE_NAME;
        pathData.setSummary(path);

        Preference backupData = findPreference(KEY_PREF_DATA_BACKUP);
        backupData.setOnPreferenceClickListener(preference -> {
            viewModel.backupDatabase();
            isBackupRequested = true;
            return false;
        });

        Preference restoreData = findPreference(KEY_PREF_DATA_RESTORE);
        restoreData.setOnPreferenceClickListener(preference -> {
            viewModel.restoreDatabase();
            isRestoreRequested = true;
            return false;
        });

        Preference privacyPolicy = findPreference(KEY_PREF_POLICY);
        privacyPolicy.setOnPreferenceClickListener(preference -> {
            startActivity(new Intent(getActivity(), PolicyActivity.class));
            return false;
        });
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        viewModel.getBackupStatus().observe(getViewLifecycleOwner(), listOfInfos -> {
            if (listOfInfos == null || listOfInfos.isEmpty() || !isBackupRequested) {
                return;
            }
            WorkInfo workInfo = listOfInfos.get(0);
            if (workInfo.getState() == WorkInfo.State.SUCCEEDED) {
                SnackbarUtils.showSnackbar(view, getString(R.string.message_data_backup));
                isBackupRequested = false;
            }
        });

        viewModel.getRestoreStatus().observe(getViewLifecycleOwner(), listOfInfos -> {
            if (listOfInfos == null || listOfInfos.isEmpty() || !isRestoreRequested) {
                return;
            }
            WorkInfo workInfo = listOfInfos.get(0);
            if (workInfo.getState() == WorkInfo.State.SUCCEEDED) {
                SnackbarUtils.showSnackbar(view, getString(R.string.message_data_restore));
                isRestoreRequested = false;
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQ_WRITE_STORAGE) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                // TODO: check (finish?)
                requireActivity().finish();
                Intent intent = new Intent(getContext(), MainActivity.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK);
                startActivity(intent);
                Toast.makeText(requireActivity(), "Now app can write", Toast.LENGTH_SHORT).show();
            } else {
                Toast.makeText(requireActivity(), "The app is not allowed to write to your storage",
                        Toast.LENGTH_SHORT).show();
            }
        }
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (key.equals(Consts.KEY_GENDER) ||
                key.equals(Consts.KEY_ACTIVITY) || key.equals(Consts.KEY_PURPOSE)) {
            Preference listPref = findPreference(key);
            listPref.setSummary("%s");
            NutritionProgramUtils.setToDefault();
        }
        if (key.equals(Consts.KEY_PREF_UI_DEFAULT_TAB)) {
            Preference listPref = findPreference(key);
            listPref.setSummary("%s");
        }
        if (key.equals(KEY_PREF_DATA_LANGUAGE)) {
            Intent restartIntent = requireActivity().getPackageManager()
                    .getLaunchIntentForPackage(requireActivity().getPackageName() );
            PendingIntent intent = PendingIntent.getActivity(
                    requireActivity(), 0,
                    restartIntent, PendingIntent.FLAG_CANCEL_CURRENT);
            AlarmManager manager = (AlarmManager) requireActivity().getSystemService(Context.ALARM_SERVICE);
            manager.set(AlarmManager.RTC, System.currentTimeMillis() + 10, intent);
            System.exit(2);
        }
        if (key.equals(Consts.KEY_BIRTHDAY)) {
            Preference birthdayPref = findPreference(key);
            date.setTime(sharedPreferences.getLong(key, 0L));

            birthdayPref.setSummary(dateFormat.format(date));
            NutritionProgramUtils.setToDefault();
        }
        if (key.equals(Consts.KEY_HEIGHT)) {
            Preference heightPref = findPreference(key);
            heightPref.setSummary(sharedPreferences.getString(key, ""));
            NutritionProgramUtils.setToDefault();
        }

        if (key.equals(Consts.KEY_PROGRAM_CAL)) {
            Preference caloriesPref = findPreference(key);
            caloriesPref.setSummary(
                    String.valueOf((int) sharedPreferences.getFloat(key, 1f))
            );
            NutritionProgramUtils.update();
        }

        if (key.equals(Consts.KEY_PROGRAM_WATER)) {
            Preference waterPref = findPreference(key);
            waterPref.setSummary(
                    String.valueOf(sharedPreferences.getInt(key, NutritionProgramUtils.getDefaultWater()))
            );
        }
    }

    @Override
    public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
        birthday.set(year,monthOfYear,dayOfMonth);
        PreferenceHelper.setValue(Consts.KEY_BIRTHDAY, birthday.getTimeInMillis());
    }

    private void showDateDialog() {
        birthday.setTimeInMillis(PreferenceHelper.getValue(Consts.KEY_BIRTHDAY, Long.class, 0L));

        new DatePickerDialog(requireActivity(),this,
                birthday.get(Calendar.YEAR),
                birthday.get(Calendar.MONTH),
                birthday.get(Calendar.DAY_OF_MONTH)).show();

    }
}

File: DiaryNutrition/app/src/main/java/ediger/diarynutrition/settings/SettingsFragment.kt
package ediger.diarynutrition.settings

import android.app.AlarmManager
import android.app.DatePickerDialog
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Bundle
import android.os.Environment
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.os.LocaleListCompat
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.navigation.fragment.NavHostFragment
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.work.WorkInfo
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.snackbar.Snackbar
import ediger.diarynutrition.DATABASE_NAME
import ediger.diarynutrition.DEFAULT_LANGUAGE
import ediger.diarynutrition.KEY_ACTIVITY
import ediger.diarynutrition.KEY_BIRTHDAY
import ediger.diarynutrition.KEY_GENDER
import ediger.diarynutrition.KEY_HEIGHT
import ediger.diarynutrition.KEY_LANGUAGE
import ediger.diarynutrition.KEY_LANGUAGE_DB
import ediger.diarynutrition.KEY_PREF_UI_DEFAULT_TAB
import ediger.diarynutrition.KEY_PROGRAM_CAL
import ediger.diarynutrition.KEY_PROGRAM_WATER
import ediger.diarynutrition.KEY_PURPOSE
import ediger.diarynutrition.MainActivity
import ediger.diarynutrition.PreferenceHelper
import ediger.diarynutrition.R
import ediger.diarynutrition.intro.PolicyActivity
import ediger.diarynutrition.util.NutritionProgramUtils
import ediger.diarynutrition.util.SnackbarUtils
import ediger.diarynutrition.util.setupSnackbar
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

class SettingsFragment : PreferenceFragmentCompat(), OnSharedPreferenceChangeListener {
    private val viewModel: SettingsViewModel by viewModels()

    private val dateFormat = SimpleDateFormat("d MMMM yyyy", Locale.getDefault())
    private val date = Date()

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences, rootKey)

        PreferenceHelper.getPreferences().registerOnSharedPreferenceChangeListener(this)

        findPreference(KEY_HEIGHT).summary =
            PreferenceHelper.getValue(KEY_HEIGHT, String::class.java, "")

        with(findPreference(KEY_BIRTHDAY)) {
            date.time = PreferenceHelper.getValue(KEY_BIRTHDAY, Long::class.javaObjectType, 0L)
            summary = dateFormat.format(date)
            onPreferenceClickListener =
                Preference.OnPreferenceClickListener {
                    showDateDialog()
                    false
                }
        }

        with(findPreference(KEY_PROGRAM_CAL)) {
            val cal = PreferenceHelper.getValue(KEY_PROGRAM_CAL, Float::class.javaObjectType, 0f)
            summary = cal.toInt().toString()
            onPreferenceClickListener = Preference.OnPreferenceClickListener {
                    ChangeCaloriesDialog().show(parentFragmentManager, null)
                    false
                }
        }

        with(findPreference(KEY_PROGRAM_WATER)) {
            summary = PreferenceHelper.getValue(
                KEY_PROGRAM_WATER,
                Int::class.javaObjectType,
                NutritionProgramUtils.getDefaultWater()
            ).toString()
            onPreferenceClickListener = Preference.OnPreferenceClickListener {
                    ChangeWaterDialog().show(parentFragmentManager, null)
                    false
                }
        }

        with(findPreference(KEY_PROGRAM_WATER)) {
            summary = PreferenceHelper.getValue(
                KEY_PROGRAM_WATER,
                Int::class.javaObjectType,
                NutritionProgramUtils.getDefaultWater()
            ).toString()
            onPreferenceClickListener = Preference.OnPreferenceClickListener {
                    ChangeWaterDialog().show(parentFragmentManager, null)
                    false
                }
        }

        findPreference(KEY_PREF_PROGRAM_EDIT).onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                startActivity(Intent(requireActivity(), ProgramActivity::class.java))
                false
            }

        findPreference(KEY_PREF_PROGRAM_MEALS).onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                NavHostFragment.findNavController(this)
                    .navigate(R.id.action_setting_to_user_meal)
                false
            }
        findPreference(KEY_PREF_PROGRAM_RESET).onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                NutritionProgramUtils.setToDefault()
                false
            }

        findPreference(KEY_PREF_DATA_BACKUP).onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                viewModel.backupDatabase()
                false
            }

        findPreference(KEY_PREF_DATA_RESTORE).onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                MaterialAlertDialogBuilder(requireActivity())
                    .setTitle(getString(R.string.dialog_restore_title))
                    .setMessage(getString(R.string.dialog_restore_message))
                    .setNegativeButton(getString(R.string.dialog_cancel)) { _, _ -> }
                    .setPositiveButton(getString(R.string.dialog_restore_positive)) { _, _ ->
                        viewModel.restoreDatabase()
                    }
                    .show()
                false
            }

        findPreference(KEY_PREF_POLICY).onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                startActivity(Intent(requireActivity(), PolicyActivity::class.java))
                false
            }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view.setupSnackbar(viewLifecycleOwner, viewModel.snackbarMessage, Snackbar.LENGTH_SHORT)

        viewModel.backupStatus.observe(viewLifecycleOwner) { listOfInfos: List<WorkInfo>? ->
            listOfInfos?.let { viewModel.concludeBackup(listOfInfos) }
        }
        viewModel.restoreStatus.observe(viewLifecycleOwner) { listOfInfos: List<WorkInfo>? ->
            listOfInfos?.let { viewModel.concludeRestore(listOfInfos) }
        }

        val toggleProgressAction = Observer<Boolean> { toggle ->
            (requireActivity() as MainActivity).toggleProgress(toggle)
        }
        viewModel.isBackupRequested.observe(viewLifecycleOwner, toggleProgressAction)
        viewModel.isRestoreRequested.observe(viewLifecycleOwner, toggleProgressAction)

    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        when (key) {
            KEY_GENDER, KEY_ACTIVITY, KEY_PURPOSE -> {
                findPreference(key).summary = "%s"
                NutritionProgramUtils.setToDefault()
            }
            KEY_PREF_UI_DEFAULT_TAB -> findPreference(key).summary = "%s"
            KEY_LANGUAGE -> {
                sharedPreferences.getString(key, DEFAULT_LANGUAGE)?.let { language ->
                    val appLocale: LocaleListCompat = LocaleListCompat.forLanguageTags(language)
                    AppCompatDelegate.setApplicationLocales(appLocale)
                }
            }
            KEY_LANGUAGE_DB -> {
                sharedPreferences.getString(key, DEFAULT_LANGUAGE)?.let { language ->
                    viewModel.changeDbLanguage(language)
                }
            }
            KEY_BIRTHDAY -> {
                date.time = sharedPreferences.getLong(key, 0L)
                findPreference(key).summary = dateFormat.format(date)
                NutritionProgramUtils.setToDefault()
            }
            KEY_HEIGHT -> {
                findPreference(key).summary = sharedPreferences.getString(key, "")
                NutritionProgramUtils.setToDefault()
            }
            KEY_PROGRAM_CAL -> {
                findPreference(key).summary = sharedPreferences.getFloat(key, 1f).toInt().toString()
                NutritionProgramUtils.update()
            }
            KEY_PROGRAM_WATER -> findPreference(key).summary =
                sharedPreferences.getInt(key, NutritionProgramUtils.getDefaultWater()).toString()
        }
    }

    private fun showDateDialog() {
        val birthday = Calendar.getInstance().apply {
            timeInMillis = PreferenceHelper.getValue(KEY_BIRTHDAY, Long::class.javaObjectType, 0L)
        }
        DatePickerDialog(
            requireActivity(),
            { _, year, monthOfYear, dayOfMonth ->
                birthday[year, monthOfYear] = dayOfMonth
                PreferenceHelper.setValue(KEY_BIRTHDAY, birthday.timeInMillis)
            },
            birthday[Calendar.YEAR],
            birthday[Calendar.MONTH],
            birthday[Calendar.DAY_OF_MONTH]
        ).show()
    }

    companion object {
        private const val KEY_PREF_DATA_LANGUAGE = "data_language"
        private const val KEY_PREF_DATA_BACKUP = "data_backup"
        private const val KEY_PREF_DATA_RESTORE = "data_restore"
        private const val KEY_PREF_POLICY = "policy"
        private const val KEY_PREF_PROGRAM_EDIT = "program_edit"
        private const val KEY_PREF_PROGRAM_RESET = "program_reset"
        private const val KEY_PREF_PROGRAM_MEALS = "program_meals"
    }
}